services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java

```diff
@@ -276,18 +311,20 @@ protected Response buildErrorResponse(boolean isPostBinding, String destination,
         }
     }
 
-    private JBossSAMLURIConstants translateErrorToSAMLStatus(Error error) {
+    private SAMLError translateErrorToSAMLStatus(Error error) {
         switch (error) {
         case CANCELLED_BY_USER:
         case CANCELLED_AIA:
         case CONSENT_DENIED:
-            return JBossSAMLURIConstants.STATUS_REQUEST_DENIED;
+            return new SAMLError(JBossSAMLURIConstants.STATUS_REQUEST_DENIED, null);
         case PASSIVE_INTERACTION_REQUIRED:
         case PASSIVE_LOGIN_REQUIRED:
-            return JBossSAMLURIConstants.STATUS_NO_PASSIVE;
+            return new SAMLError(JBossSAMLURIConstants.STATUS_NO_PASSIVE, null);
+        case ALREADY_LOGGED_IN:
+            return new SAMLError(JBossSAMLURIConstants.STATUS_AUTHNFAILED, Constants.AUTHENTICATION_EXPIRED_MESSAGE);
```

Comment:
Maybe we also need to retry with this status in the saml adapter like in the JS adapter for OIDC. But this can be a follow-up or even not added at all. I have tested and the result with current adapter is you receive an unauthorized error in the app (the saml adapter in EAP receives the error, it works if you retry).

Reply:
Good point! Added https://github.com/keycloak/keycloak/issues/28412 as a follow-up task.